### ISSUE TITLE
TOC extension marker '[//]: # (TOC)' problem

### ISSUE THREAD

**Issue by troyengel** [Python-Markdown/markdown#667]
Disclaimer: no idea what I'm doing :)

I'm trying to hide the `[TOC]` marker tag in the raw Markdown file using this recipe: https://stackoverflow.com/a/20885980/150649 (which I do confirm gets hidden when viewing via Github, etc.)

Using pdb I can confirm the self.marker is getting set properly to `[//]: # (TOC)` when passed in, but it's simply being... skipped? in the output. I tried to break it down into the simplest form:

```
>>> import markdown
>>> text = '''
... [//]: # (TOC)
... # 111
... ## 222'''
>>> md = markdown.Markdown(extensions=['toc(marker=[//]: # (TOC)'])
>>> print(md.convert(text))
<h1 id="111">111</h1>
<h2 id="222">222</h2>
```

Am I doing it wrong, or is this just a sequence of marker too complicated for the extension to handle maybe? I tried tracing the code but this isn't my forte, got lost a bit trying to figure out where the strip was happening... thanks.

**Comment by waylan** [comment]
If you don’t want to include the marker, then just leave it out. Or am I miss understanding what you are trying to do?

**Comment by troyengel** [comment]
If you leave out the marker, then no TOC is generated so I'm not following your response? When you write a Markdown file and commit it to GitLab / GitHub / etc., the tag shows up in the output as raw `[TOC]` in your document when it's rendered (well, this isn't rendered - just raw text) and looks janky. it's undesirable and easy to see if you just make any random MD file here on Github.

The recipe I linked to above is how to use valid Markdown syntax to "hide" a comment (as you might do in HTML with `<!-- some comment -->`) so that it is not rendered, by using the footnote reference elements in a creative way. If you create a new MD file on Github, type in some content (and markdown) and include `[//]: # (TOC)` it does not get parsed or displayed in the output while in the git repository. It's an internal source comment only, nifty!

So the effort here is to set the Python-Markdown TOC marker tag to this special recipe, so that when passed through a generator to create HTML, it will still be found as a TOC and generate the content; however, that special recipe is hidden to Git repo views, as they all parse it into nothing visible. The problem is that the extension is just not properly parsing my tag - it sees it but strips it somehow as shown in my example. I would like the TOC extension to properly recognize this tag as the TOC marker and do the right thing. :) This way the content can be viewed cleanly on Git view, as well as generating HTML output by a renderer w/TOC.

I understand this is a complex example of a TOC marker tag and may be too much for this extension to handle right now, but according to people who know way more than me (the Internet) this is pretty much the only way to hide comments in Markdown in a compatible way. I'm just lost in the Markdown TOC code trying to figure out why it isn't working correctly to help figure out what's broken - I can determine that the `self.marker` is set correctly, the TOC marker makes it into the class attribute unmolested - then it just sort of... doesn't work.

**Comment by troyengel** [comment]
OK I'm getting somewhere, I was looking in the wrong place - the problem appears to be the ReferencePreprocessor which runs before the TOC extension runs, right here:

* https://github.com/Python-Markdown/markdown/blob/master/markdown/core.py#L231

```
(Pdb) print(self.lines)
['[//]: # (TOC)', '# 111', '## 222', '', '', '']
(Pdb) n
-> self.lines = prep.run(self.lines)
(Pdb) print(prep)
<markdown.preprocessors.ReferencePreprocessor object at 0x7f74343199b0>
(Pdb) n
-> for prep in self.preprocessors.values():
(Pdb) print(self.lines)
['', '# 111', '## 222', '', '', '']
```

...one step closer, this makes sense as I'm using a reference marker. So the marker element to replace is getting stripped out of the source before the TOC extension can replace it (verified that inside the TOC extension code).

**Comment by facelessuser** [comment]
@troyengel, that's how it is supposed to work.  Those are references that *can* be used later, but if they are not used, they don't make it to the output.  This is how it should be, and that won't change. Don't believe me, check out how other parsers handle un-linked references [here](https://johnmacfarlane.net/babelmark2/?text=Test%0A%0A%5B%2F%2F%5D%3A+%23someref).

Comments are for the person reading the source, not for the result/output. Even in coding languages, comments don't make it the binary, because they are worthless to the binary.  Now references are **not** comments, they are just used in this hacky way by some people.  But un-linked references are also worthless in the output.

Personally, if you want to comment, just use HTML comments. Just remember that it is HTML and does take up space and is parsed as HTML, so put some empty lines around it.

```
<!--
[TOC]
-->
```

If you want to output a table of contents, but not show it (which sounds like what you really want), commenting with a reference hack is **not** the answer.  You should just output the TOC and hide it via CSS by targeting the `.toc` class.

**Comment by troyengel** [comment]
Yeah, that's it - the ReferencePreprocessor is stripping out my reference tag before the TOC extension can see it by design, which sort of leaves me wondering if there's any way to solve this other than skipping this routine entirely. Inside it... (abbreviated output):

* https://github.com/Python-Markdown/markdown/blob/master/markdown/preprocessors.py#L323

```
 ... ... ...
-> line = lines.pop(0)

(Pdb) print(line)
[//]: # (TOC)

(Pdb) print(self.RE)
re.compile('^[ ]{0,3}\\[([^\\]]*)\\]:\\s*([^ ]*)[ ]*([ ]*(\\"(.*)\\"|\\\'(.*)\\\'|\\((.*)\\))[ ]*)?$', re.DOTALL)

-> m = self.RE.match(line)
-> if m:
-> id = m.group(1).strip().lower()
-> link = m.group(2).lstrip('<').rstrip('>')
-> t = m.group(5) or m.group(6) or m.group(7)
-> if not t:
-> self.markdown.references[id] = (link, t)

(Pdb) print(self.markdown.references['//'])
('#', 'TOC')

-> return new_text  # + "\n"
(Pdb) print(new_text)
['', '# 111', '## 222', '', '', '']
```

**Comment by facelessuser** [comment]
Please see my previous comment as I believe it answers your question.

**Comment by troyengel** [comment]
@facelessuser that's not what I want at all - I've discovered the problem as outlined above, it's an order of operations. The reference preprocessor is stripping and stashing my reference tag *before* it's passed to the TOC extension processor to see it. Nobody is saying this is improper or not as designed (not sure why you're getting all aggro), it's just how Python-Markdown is designed in code flow now that I have debugged it. Now that I understand why it doesn't work how *I* want it to, I can develop a workaround.

**Comment by facelessuser** [comment]
I'm not sure what you mean by *aggro*, aggressive maybe? but what I was doing was laying out that fundamentally, this is how **all** Markdown processors work.  I don't think I was being aggressive, but instead was trying to be informative. Maybe I was too direct? Anyways, I also wanted to be clear that changing this behavior isn't going to happen (if there was intention for it to be officially changed).

I also pointed out the most sensible way to hide the toc and still output it (within the existing default options and features).

> If you want to output a table of contents, but not show it (which sounds like what you really want), commenting with a reference hack is not the answer. You should just output the TOC and hide it via CSS by targeting the `.toc` class.

If this approach is not to your liking, and you are really set on using this reference approach, then feel free to continue, but there will be more issues as well.

Even if you changed the "flow" and toc is expanded in the references, un-linked references are still thrown away, which is what I was also illustrating. Modifying references to allow TOC expansion will not solve your problem.

You may just want to write your own commenting extension, or your own custom reference extension (and in this case replace the official one when it loads). I guess you are also free to fork the official repo and modify it to your liking as well. Another option is to just write a post processor that looks for TOC and hides it, if that is your desire.

 Maybe I've also misunderstood what you are trying to accomplish?

**Comment by troyengel** [comment]
Yep, you've misunderstood and have it backwards. Also, the HTML comment doesn't work because there's a regex substitution in the TOC extension itself which strips out all HTML entities before processing the comment (see `stashedHTML2text(text, md)`) so that's a red herring as well, that ws the first thing I tried. Closing, this will go nowhere now that I have debugged why it won't work with this code.
